pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -1369,25 +1369,26 @@ func (c *Controller) getProxyServiceInstancesByPod(pod *v1.Pod,
 }
 
 func (c *Controller) GetProxyWorkloadLabels(proxy *model.Proxy) labels.Instance {
+	var out labels.Instance
 	pod := c.pods.getPodByProxy(proxy)
 	if pod != nil {
 		if _, exist := pod.Labels[model.LocalityLabel]; exist {
-			return pod.Labels
-		}
-		locality := c.getPodLocality(pod)
-		if locality == "" {
-			return pod.Labels
-		}
-		out := make(map[string]string, len(pod.Labels)+1)
-		for k, v := range pod.Labels {
-			out[k] = v
+			out = pod.Labels
+		} else if locality := c.getPodLocality(pod); locality == "" {
+			out = pod.Labels
+		} else {
+			out = make(map[string]string, len(pod.Labels)+1)
+			for k, v := range pod.Labels {
+				out[k] = v
+			}
+			// Add locality labels to support locality Load balancing for proxy without service instances.
+			// As this may contain node topology labels, which could not be got from aggregator controller
+			out[model.LocalityLabel] = locality
 		}
-		// Add locality labels to support locality Load balancing for proxy without service instances.
-		// As this may contain node topology labels, which could not be got from aggregator controller
-		out[model.LocalityLabel] = locality
-		return out
+		// set k8s node name label, for ServiceInternalTrafficPolicy
```

Comment:
I think this should be in AugmentLabels

Reply:
I'm not sure how to do that. We have to store the k8s node somewhere, is there any other place for the kube controller to set the k8s node for the proxy?

Reply:
For endpoint labels we can use `AugmentLabels` (updated), but for proxy labels I'm not sure how, is there any other place to get the k8s node?

Reply:
This is mutating the underlying pod labels

Reply:
can set it in `setTopologyLabels`

Reply:
> This is mutating the underlying pod labels

Thanks for reminding, fixed

> can set it in `setTopologyLabels`

I think we have to set the host label here, as we can't get it elsewhere

Reply:
Why not AugmentLabels then fetch it like `setTopologyLabels` ( `proxy.ServiceInstances[0].Endpoint.Locality.Node`)

Reply:
Actually we can add NODE_NAME to the injection template (downward API) and get it from there 99% of the time. And fallback to ^ when its not there.

Reply:
The experimental-ambient branch already does this if you would like a reference